mds/MDLog: a large refactor

leonid-s-usov opened the pull request:
* Make sure that any structure referring to a LogSegment is using a shared pointer to do that
* Limit the API changes by enhancing the LogSegment class with an `enable_shared_from_this` base
  - Make LogSegment::LogSegment private and expose a `create` method to correctly initialize the base class
  - Introduce AutoSharedLogSegment helper class that recovers a shared pointer from a LogSegment*
* Make the needed changes to MDRank and MDLog to make MDLog unit-testable
* Large refactoring of the MDLog post the LogSegment changes

Fixes: https://tracker.ceph.com/issues/4744

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

batrick commented on src/mds/LogEvent.h:
```diff
@@ -49,10 +49,12 @@
 class MDSRank;
 class LogSegment;
 class EMetaBlob;
+using AutoSharedLogSegment = auto_shared_ptr<LogSegment>;
```
So I'm mystified by the utility of this class. Why couldn't it be:

`using foo = std::shared_ptr<LogSegment>`

?

leonid-s-usov replied:
As I mentioned in the commit message, the utility of this class is to automatically convert a pointer to an object that inherits from [`enable_shared_from_this`](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this) into a shared pointer and also adds an implicit conversion from the shared pointer back to a plain pointer. This helps minimize refactoring.

I'll elaborate on how I believe this helps address the original ticket in a separate comment.

batrick replied:
Normally we use `boost::intrusive_ptr` if other accessors have a raw pointer. That's somewhat inappropriate in a code base we fully control, the mds.

Your autoshared class is interesting though. I'll have to think about this more.

batrick commented on src/mds/LogSegment.h:
```diff
@@ -50,6 +51,13 @@ class LogSegment {
+
+  [[nodiscard]] static AutoSharedLogSegment create(uint64_t _seq, loff_t off = -1)
+  {
+    return std::shared_ptr<LogSegment>(new LogSegment(_seq, off));
```
https://en.cppreference.com/w/cpp/memory/shared_ptr/make_shared

leonid-s-usov replied:
`make_shared` is inaccessible here because the constructor is private. See class `Best()` in the example from the [documentation on shared_ptr](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this)

Additionally, `make_shared` comes at the cost of pinning the memory of the whole object when only weak pointers are left since the same memory block is used to hold the shared pointer control and the class data. This is mentioned in the second note on the documentation page you've referenced.

batrick replied:
> `make_shared` is inaccessible here because the constructor is private. See class `Best()` in the example from the [documentation on shared_ptr](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this)

ok

> Additionally, `make_shared` comes at the cost of pinning the memory of the whole object when only weak pointers are left since the same memory block is used to hold the shared pointer control and the class data. This is mentioned in the second note on the documentation page you've referenced. To clarify, I'm not sure this will be a blocker for us, and maybe we could even benefit from it, but I think it's an important point to keep in mind

I don't really think that will be significant but I also wonder if a clever implementation is permitted to `realloc` the memory block when `T` is destroyed. What an odd caveat I had the opposite understanding about.

batrick reviewed (changes requested):
My main concern with this PR is that it doesn't actually address:

"These really ought to be ref-counted in some way to prevent early expiry."

from https://tracker.ceph.com/issues/4744

It's not well laid out what the concern is but I suspect it's something like to prevent tickets like:

https://tracker.ceph.com/issues/59119

You could potentially resolve that ticket with this PR by checking that the `LogSegment` reference count is 1 before trimming* the segment.

* I believe Greg actually meant "early trimming" as that's the real concern: that the log segment will be removed from the physical journal.

cc @gregsfortytwo

batrick commented:
> > My main concern with this PR is that it doesn't actually address:
> > "These really ought to be ref-counted in some way to prevent early expiry."
> 
> @batrick I do think that this PR addresses the original ticket. It does so by changing all containers to hold a shared pointer to the LogSegment instead of a plain pointer.

You've refcounted `LogSegment` but you haven't used that **to prevent early expiry**. Specifically, to address that I think you need to change:

https://github.com/ceph/ceph/blob/ce9771bbebacbdaa00156429bc382aaa4d40ea61/src/mds/MDLog.cc#L842-L851

to not trim (i.e. delete) a segment unless its refcount is 1.

batrick commented on src/mds/LogSegment.h:
```diff
@@ -34,10 +35,10 @@ class CDentry;
 struct MDPeerUpdate;
 
-class LogSegment {
- public:
```
`public:` starts a class. See:

https://google.github.io/styleguide/cppguide.html#Declaration_Order

leonid-s-usov replied:
It says
> A class definition should **usually** start with a public: section, ...
(the emphasis is mine)

Also, below it requests that constructors precede other members. 
So, I'm not convinced this violates the requirement, but I will move it to the end if you think I should, just confirm please.

batrick replied:
Within `mds/` I took that "request" as policy because the status-quo was a complete mess of public/private back-n-forth. Cleaning that up used valuable dev time so I'm a stickler about not regressing.  See the commit history of the header files if interested.

leonid-s-usov replied:
done

leonid-s-usov commented:
> to not trim (i.e. delete) a segment unless its refcount is 1.

Well, I don't need to do anything for the log segment to be deleted only when the last strong reference to it is gone, that's the primary function of a shared pointer. In other words, if the delete operation is what we'd like to delay - that's already the case with the current implementation.

gregsfortytwo commented:
Specific questions to look at here:
1) the locking changes around the submit lock, and taking the mds lock in the thread
2) ambiguity around event counts (live versus expiring versus expired) versus health checks

batrick commented on src/mds/MDLog.cc:
```diff
@@ -416,65 +416,72 @@ void MDLog::_submit_thread()
-	logger->set(l_mdl_wrpos, ls->end);
+      // acquire the mds lock until the end of the scope
+      // to update the LS offsets and major offsets
+      std::lock_guard l(mds->mds_lock);
```
I still need to go through these changes in more detail but this will not fly. The `mds_lock` has too much thread contention already. Also, the submit thread is intended to operate completely outside the `mds_lock`. I do not think we should ever change that.

I believe there may be another way to protect the LogSegment members behind the `mds_lock`: add the `LogSegment` to `C_MDL_Flushed` and update the segment's end/offset in the completion (`::finish`).

leonid-s-usov replied:
C_MDL_Flushed will take mds_lock, so not much difference in terms of contention. But I agree that we could benefit from performing the update on the finisher thread instead of the submit thread.

batrick replied:
As you said, the journaler thread already regularly acquires the `mds_lock` so it's advantageous to use it for this purpose instead of adding a new thread to the contention.

leonid-s-usov replied:
Please see a1c9d1a8be08807d83eac0c7bf75cfe04101768c, I removed the need for locking.

leonid-s-usov commented:
@batrick please see the fixup 53834427f05141b5289f3b0668ef0c107252ed7e where I reverted my previous bad change to `C_MDS_VoidFn` and introduced a new wrapper that acquires the mds lock before calling into a wrapped MDSInternalContext.

leonid-s-usov commented:
jenkins test make check

gregsfortytwo commented on src/mds/MDSRank.h:
```diff
@@ -228,7 +392,7 @@ class MDSRank {
 
     void handle_write_error(int err);
-    void handle_write_error_with_lock(int err);
+    void handle_write_error_unlocked(int err);
```
Can you rename functions as a separate commit? I happened to notice it while we were talking and then had to scramble to figure out why you were changing the lock state, only to realize you had renamed the function (presumably you found it misleading?).

gregsfortytwo replied:
Yes please -- it makes the correctness checks way easier if renames are independent of anything else. :)

leonid-s-usov commented:
jenkins test make check

vshankar commented:
jenkins retest this please

leonid-s-usov commented:
@vshankar apparently, there are some mdlog unit test timeouts that I need to investigate. I will get back to it after the quiesce db

leonid-s-usov commented:
jenkins test make check

leonid-s-usov commented:
jenkins test make check arm64

leonid-s-usov commented:
https://pulpito.ceph.com/leonidus-2024-06-25_18:06:42-fs-wip-lusov-mdlog-distro-default-smithi/
This run has one failure, caused by the zombie segment detection, new in this PR. It means a log segment is still kept somewhere as a strong reference despite having expired. To be investigated.

leonid-s-usov commented:
> https://pulpito.ceph.com/leonidus-2024-06-25_18:06:42-fs-wip-lusov-mdlog-distro-default-smithi/ This run has one failure, caused by the zombie segment detection, new in this PR. It means a log segment is still kept somewhere as a strong reference despite having expired. To be investigated.

There are 2 more failures. xfstests-dev, which I am re-running now to see if this is easily reproduced, and a crash in the trimming. That one I'll address today

leonid-s-usov commented:
> https://pulpito.ceph.com/leonidus-2024-06-25_18:06:42-fs-wip-lusov-mdlog-distro-default-smithi/ This run has one failure, caused by the zombie segment detection, new in this PR. It means a log segment is still kept somewhere as a strong reference despite having expired. To be investigated.

Correction:
there are two failures related to the zombie detection which is enabled by the new dedicated dev setting: `mds_debug_zombie_log_segments`

- https://pulpito.ceph.com/leonidus-2024-06-25_18:06:42-fs-wip-lusov-mdlog-distro-default-smithi/7771770
- https://pulpito.ceph.com/leonidus-2024-06-25_18:06:42-fs-wip-lusov-mdlog-distro-default-smithi/7771779

There are also 16 xfs tailures here:
https://pulpito.ceph.com/leonidus-2024-06-25_18:06:42-fs-wip-lusov-mdlog-distro-default-smithi/7771785/

leonid-s-usov commented:
I'm running 10 more instances of the xfs tests here: https://pulpito.ceph.com/leonidus-2024-06-26_05:12:19-fs-wip-lusov-mdlog-distro-default-smithi/

leonid-s-usov commented:
@vshankar known issues at this point:

* the new zombie detection setting (see above) is not supported in upgrade workloads because the previous version doesn't recognize the setting. I believe there's a standard approach to dealing with this kind of problem when new settings are introduced
* with the zombie detection enabled, some tests cause MDS crashes with `FAILED ceph_assert(ignore_zombies)`. It could be a real leak, but it could also be some internal problem where a strong ref is kept somewhere in the MDLog itself.
* Not yet sure if the XFS test failures are related to this PR

I haven't scheduled a full fs suite on this change because it is not even started with priority 100.
